File: ios/ExpoBraintreeModule.swift
import ExpoModulesCore
import Braintree
import Contacts
import PassKit

public class ExpoBraintreeModule: Module {
  private var authorization: String?

  public func definition() -> ModuleDefinition {
    Name("ExpoBraintree")

    // MARK: - Initialization

    AsyncFunction("initialize") { (auth: String) in
      guard !auth.isEmpty else {
        throw BraintreeError.initializationFailed
      }
      self.authorization = auth
    }

    // MARK: - Card Tokenization

    AsyncFunction("tokenizeCard") { (cardData: CardData) -> [String: Any?] in
      let auth = try self.requireAuthorization()
      let cardClient = BTCardClient(authorization: auth)

      let card = BTCard(
        number: cardData.number,
        expirationMonth: cardData.expirationMonth,
        expirationYear: cardData.expirationYear,
        cvv: cardData.cvv ?? "",
        postalCode: cardData.postalCode,
        cardholderName: cardData.cardholderName
      )

      let nonce = try await cardClient.tokenize(card)

      return [
        "nonce": nonce.nonce,
        "type": "card",
        "isDefault": nonce.isDefault,
        "description": nonce.description,
        "cardNetwork": nonce.cardNetwork.humanReadable,
        "lastFour": nonce.lastFour,
        "expirationMonth": nonce.expirationMonth,
        "expirationYear": nonce.expirationYear,
        "bin": nonce.bin,
      ]
    }

    // MARK: - Apple Pay

    AsyncFunction("isApplePaySupported") { () -> Bool in
      return PKPaymentAuthorizationController.canMakePayments()
    }

    AsyncFunction("tokenizeApplePay") { (request: ApplePayRequestData) -> [String: Any?] in
      let auth = try self.requireAuthorization()
      let applePayClient = BTApplePayClient(authorization: auth)

      let paymentRequest = try await applePayClient.makePaymentRequest()
      paymentRequest.merchantIdentifier = request.merchantIdentifier
      paymentRequest.countryCode = request.countryCode
      paymentRequest.currencyCode = request.currencyCode
      paymentRequest.merchantCapabilities = .capability3DS
      paymentRequest.paymentSummaryItems = request.paymentSummaryItems.map { item in
        PKPaymentSummaryItem(
          label: item.label,
          amount: NSDecimalNumber(string: item.amount),
          type: item.type == "pending" ? .pending : .final
        )
      }

      if let shippingFields = request.requiredShippingContactFields {
        paymentRequest.requiredShippingContactFields = Set(shippingFields.compactMap { field in
          switch field {
          case "postalAddress": return .postalAddress
          case "name": return .name
          case "emailAddress": return .emailAddress
          case "phoneNumber": return .phoneNumber
          default: return nil
          }
        })
      }

      if let networks = request.supportedNetworks {
        paymentRequest.supportedNetworks = networks.compactMap { network in
          switch network {
          case "visa": return .visa
          case "masterCard": return .masterCard
          case "amex": return .amex
          case "discover": return .discover
          case "maestro": return .maestro
          case "jcb": return .JCB
          default: return nil
          }
        }
      }

      return try await withCheckedThrowingContinuation { continuation in
        let controller = PKPaymentAuthorizationController(paymentRequest: paymentRequest)
        let delegate = ApplePayDelegate(applePayClient: applePayClient, continuation: continuation)

        // Prevent delegate from being deallocated
        objc_setAssociatedObject(controller, "delegate", delegate, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)

        controller.delegate = delegate
        Task {
          let presented = await controller.present()
          if !presented {
            continuation.resume(throwing: BraintreeError.applePayPresentationFailed)
          }
        }
      }
    }

    // MARK: - PayPal Checkout

    AsyncFunction("tokenizePayPalCheckout") { (request: PayPalCheckoutRequestData) -> [String: Any?] in
      let auth = try self.requireAuthorization()
      let paypalClient = await MainActor.run { BTPayPalClient(authorization: auth) }

      let intent: BTPayPalRequestIntent
      switch request.intent {
      case "sale": intent = .sale
      case "order": intent = .order
      default: intent = .authorize
      }

      let checkoutRequest = BTPayPalCheckoutRequest(
        amount: request.amount,
        intent: intent,
        userAction: request.userAction == "commit" ? .payNow : .none,
        currencyCode: request.currencyCode,
        displayName: request.displayName,
        isShippingAddressEditable: request.shippingAddressEditable ?? false,
        isShippingAddressRequired: request.shippingAddressRequired ?? false
      )

      let nonce = try await paypalClient.tokenize(checkoutRequest)
      return Self.serializePayPalNonce(nonce)
    }

    // MARK: - PayPal Vault

    AsyncFunction("tokenizePayPalVault") { (request: PayPalVaultRequestData) -> [String: Any?] in
      let auth = try self.requireAuthorization()
      let paypalClient = await MainActor.run { BTPayPalClient(authorization: auth) }

      let vaultRequest = BTPayPalVaultRequest(
        billingAgreementDescription: request.billingAgreementDescription,
        displayName: request.displayName,
        isShippingAddressEditable: request.shippingAddressEditable ?? false,
        isShippingAddressRequired: request.shippingAddressRequired ?? false
      )

      let nonce = try await paypalClient.tokenize(vaultRequest)
      return Self.serializePayPalNonce(nonce)
    }

    // MARK: - Venmo

    AsyncFunction("tokenizeVenmo") { (request: VenmoRequestData) -> [String: Any?] in
      let auth = try self.requireAuthorization()

      guard let universalLink = URL(string: request.universalLink) else {
        throw BraintreeError.invalidUniversalLink
      }

      let venmoClient = BTVenmoClient(authorization: auth, universalLink: universalLink)

      let venmoRequest = BTVenmoRequest(
        paymentMethodUsage: request.paymentMethodUsage == "multiUse" ? .multiUse : .singleUse,
        profileID: request.profileId,
        displayName: request.displayName,
        collectCustomerBillingAddress: request.collectCustomerBillingAddress ?? false,
        collectCustomerShippingAddress: request.collectCustomerShippingAddress ?? false
      )

      let nonce = try await venmoClient.tokenize(venmoRequest)

      return [
        "nonce": nonce.nonce,
        "type": "venmo",
        "isDefault": nonce.isDefault,
        "description": nonce.description,
        "username": nonce.username,
        "billingAddress": Self.serializePostalAddress(nonce.billingAddress),
        "shippingAddress": Self.serializePostalAddress(nonce.shippingAddress),
      ]
    }
  }

  // MARK: - Helpers

  private func requireAuthorization() throws -> String {
    guard let auth = authorization else {
      throw BraintreeError.notInitialized
    }
    return auth
  }

  private static func serializePayPalNonce(_ nonce: BTPayPalAccountNonce) -> [String: Any?] {
    return [
      "nonce": nonce.nonce,
      "type": "paypal",
      "isDefault": nonce.isDefault,
      "description": nonce.description,
      "email": nonce.email,
      "firstName": nonce.firstName,
      "lastName": nonce.lastName,
      "billingAddress": serializePostalAddress(nonce.billingAddress),
      "shippingAddress": serializePostalAddress(nonce.shippingAddress),
    ]
  }

  private static func serializePostalAddress(_ address: BTPostalAddress?) -> [String: Any?]? {
    guard let address = address else { return nil }
    let components = address.addressComponents()
    return [
      "recipientName": components["recipientName"],
      "streetAddress": components["streetAddress"],
      "extendedAddress": components["extendedAddress"],
      "locality": components["locality"],
      "region": components["region"],
      "postalCode": components["postalCode"],
      "countryCodeAlpha2": components["countryCodeAlpha2"],
    ]
  }
}

// MARK: - Record Types

struct CardData: Record {
  @Field var number: String
  @Field var expirationMonth: String
  @Field var expirationYear: String
  @Field var cvv: String?
  @Field var postalCode: String?
  @Field var cardholderName: String?
}

struct ApplePaySummaryItemData: Record {
  @Field var label: String
  @Field var amount: String
  @Field var type: String?
}

struct ApplePayRequestData: Record {
  @Field var merchantIdentifier: String
  @Field var countryCode: String
  @Field var currencyCode: String
  @Field var paymentSummaryItems: [ApplePaySummaryItemData]
  @Field var supportedNetworks: [String]?
  @Field var requiredShippingContactFields: [String]?
}

struct PayPalCheckoutRequestData: Record {
  @Field var amount: String
  @Field var currencyCode: String?
  @Field var intent: String?
  @Field var userAction: String?
  @Field var displayName: String?
  @Field var shippingAddressRequired: Bool?
  @Field var shippingAddressEditable: Bool?
}

struct PayPalVaultRequestData: Record {
  @Field var billingAgreementDescription: String?
  @Field var displayName: String?
  @Field var userAuthenticationEmail: String?
  @Field var shippingAddressRequired: Bool?
  @Field var shippingAddressEditable: Bool?
}

struct VenmoRequestData: Record {
  @Field var paymentMethodUsage: String
  @Field var universalLink: String
  @Field var profileId: String?
  @Field var displayName: String?
  @Field var collectCustomerBillingAddress: Bool?
  @Field var collectCustomerShippingAddress: Bool?
}

// MARK: - Errors

enum BraintreeError: Error, LocalizedError {
  case notInitialized
  case initializationFailed
  case applePayPresentationFailed
  case applePayCancelled
  case invalidUniversalLink

  var errorDescription: String? {
    switch self {
    case .notInitialized:
      return "Braintree client not initialized. Call initialize() first."
    case .initializationFailed:
      return "Failed to initialize Braintree client. Check your authorization token."
    case .applePayPresentationFailed:
      return "Failed to present Apple Pay payment sheet."
    case .applePayCancelled:
      return "User cancelled Apple Pay."
    case .invalidUniversalLink:
      return "Invalid universal link URL provided for Venmo."
    }
  }
}

// MARK: - Apple Pay Delegate

private class ApplePayDelegate: NSObject, PKPaymentAuthorizationControllerDelegate {
  private let applePayClient: BTApplePayClient
  private var continuation: CheckedContinuation<[String: Any?], Error>?
  private var didAuthorize = false

  init(applePayClient: BTApplePayClient, continuation: CheckedContinuation<[String: Any?], Error>) {
    self.applePayClient = applePayClient
    self.continuation = continuation
  }

  func paymentAuthorizationController(
    _ controller: PKPaymentAuthorizationController,
    didAuthorizePayment payment: PKPayment,
    handler completion: @escaping (PKPaymentAuthorizationResult) -> Void
  ) {
    didAuthorize = true
    Task {
      do {
        let nonce = try await applePayClient.tokenize(payment)
        completion(PKPaymentAuthorizationResult(status: .success, errors: nil))

        var result: [String: Any?] = [
          "nonce": nonce.nonce,
          "type": "applePay",
          "isDefault": nonce.isDefault,
          "description": nonce.description,
        ]

        if let contact = payment.shippingContact {
          if let postalAddress = contact.postalAddress {
            result["shippingAddress"] = [
              "recipientName": contact.name.map { PersonNameComponentsFormatter().string(from: $0) },
              "streetAddress": postalAddress.street,
              "extendedAddress": postalAddress.subLocality.isEmpty ? nil : postalAddress.subLocality,
              "locality": postalAddress.city,
              "region": postalAddress.state,
              "postalCode": postalAddress.postalCode,
              "countryCodeAlpha2": postalAddress.isoCountryCode,
            ] as [String: Any?]
          }
          result["email"] = contact.emailAddress
          result["phoneNumber"] = contact.phoneNumber?.stringValue
        }

        continuation?.resume(returning: result)
        continuation = nil
      } catch {
        completion(PKPaymentAuthorizationResult(status: .failure, errors: [error]))
        continuation?.resume(throwing: error)
        continuation = nil
      }
    }
  }

  func paymentAuthorizationControllerDidFinish(_ controller: PKPaymentAuthorizationController) {
    controller.dismiss {
      if !self.didAuthorize {
        self.continuation?.resume(throwing: BraintreeError.applePayCancelled)
        self.continuation = nil
      }
    }
  }
}

// MARK: - Card Network Extension

extension BTCardNetwork {
  var humanReadable: String {
    switch self {
    case .unknown: return "Unknown"
    case .AMEX: return "Amex"
    case .dinersClub: return "DinersClub"
    case .discover: return "Discover"
    case .masterCard: return "MasterCard"
    case .visa: return "Visa"
    case .JCB: return "JCB"
    case .laser: return "Laser"
    case .maestro: return "Maestro"
    case .unionPay: return "UnionPay"
    case .hiper: return "Hiper"
    case .hipercard: return "Hipercard"
    case .solo: return "Solo"
    case .`switch`: return "Switch"
    case .ukMaestro: return "UKMaestro"
    @unknown default: return "Unknown"
    }
  }
}

File: android/src/main/java/expo/modules/braintree/ExpoBraintreeModule.kt
package expo.modules.braintree

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.activity.ComponentActivity
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

// Card
import com.braintreepayments.api.card.CardClient
import com.braintreepayments.api.card.Card
import com.braintreepayments.api.card.CardResult

// Google Pay
import com.braintreepayments.api.googlepay.GooglePayClient
import com.braintreepayments.api.googlepay.GooglePayRequest
import com.braintreepayments.api.googlepay.GooglePayResult
import com.braintreepayments.api.googlepay.GooglePayCardNonce
import com.braintreepayments.api.googlepay.GooglePayLauncher
import com.braintreepayments.api.googlepay.GooglePayPaymentAuthRequest
import com.braintreepayments.api.googlepay.GooglePayPaymentAuthResult
import com.braintreepayments.api.googlepay.GooglePayReadinessResult
import com.braintreepayments.api.googlepay.GooglePayTotalPriceStatus

// PayPal
import com.braintreepayments.api.paypal.PayPalClient
import com.braintreepayments.api.paypal.PayPalCheckoutRequest
import com.braintreepayments.api.paypal.PayPalVaultRequest
import com.braintreepayments.api.paypal.PayPalAccountNonce
import com.braintreepayments.api.paypal.PayPalResult
import com.braintreepayments.api.paypal.PayPalPaymentIntent
import com.braintreepayments.api.paypal.PayPalPaymentUserAction
import com.braintreepayments.api.paypal.PayPalLauncher
import com.braintreepayments.api.paypal.PayPalPaymentAuthRequest
import com.braintreepayments.api.paypal.PayPalPaymentAuthResult
import com.braintreepayments.api.paypal.PayPalPendingRequest

// Venmo
import com.braintreepayments.api.venmo.VenmoClient
import com.braintreepayments.api.venmo.VenmoRequest
import com.braintreepayments.api.venmo.VenmoResult
import com.braintreepayments.api.venmo.VenmoAccountNonce
import com.braintreepayments.api.venmo.VenmoPaymentMethodUsage
import com.braintreepayments.api.venmo.VenmoLauncher
import com.braintreepayments.api.venmo.VenmoPaymentAuthRequest
import com.braintreepayments.api.venmo.VenmoPaymentAuthResult
import com.braintreepayments.api.venmo.VenmoPendingRequest

// Core
import com.braintreepayments.api.core.PostalAddress

class ExpoBraintreeModule : Module() {
  private var authorization: String? = null
  private var appLinkReturnUrl: Uri? = null

  // Google Pay state
  private var googlePayLauncher: GooglePayLauncher? = null
  private var googlePayClient: GooglePayClient? = null
  private var googlePayPromise: Promise? = null

  // PayPal state
  private val payPalLauncher = PayPalLauncher()
  private var payPalClient: PayPalClient? = null
  private var payPalPendingRequestString: String? = null
  private var payPalPromise: Promise? = null

  // Venmo state
  private val venmoLauncher = VenmoLauncher()
  private var venmoClient: VenmoClient? = null
  private var venmoPendingRequestString: String? = null
  private var venmoPromise: Promise? = null

  private val currentActivity
    get() = appContext.activityProvider?.currentActivity
      ?: throw BraintreeNotInitializedException()

  private val currentContext: Context
    get() = appContext.reactContext
      ?: throw BraintreeNotInitializedException()

  override fun definition() = ModuleDefinition {
    Name("ExpoBraintree")

    // Handle PayPal/Venmo return from browser/app
    OnNewIntent { intent ->
      handlePayPalReturn(intent)
      handleVenmoReturn(intent)
    }

    OnActivityEntersForeground {
      currentActivity.intent?.let { intent ->
        handlePayPalReturn(intent)
        handleVenmoReturn(intent)
      }
    }

    // ── Initialization ────────────────────────────────────────────────────

    AsyncFunction("initialize") { auth: String ->
      authorization = auth

      // Try to initialize Google Pay launcher (needs ComponentActivity)
      try {
        val activity = currentActivity
        if (activity is ComponentActivity && googlePayLauncher == null) {
          googlePayLauncher = GooglePayLauncher(activity) { paymentAuthResult ->
            handleGooglePayReturn(paymentAuthResult)
          }
        }
      } catch (_: Exception) {
        // Google Pay launcher creation may fail if called after onStart
      }
    }

    AsyncFunction("setReturnUrl") { url: String ->
      appLinkReturnUrl = Uri.parse(url)
    }

    // ── Card Tokenization ─────────────────────────────────────────────────

    AsyncFunction("tokenizeCard") { cardData: CardDataRecord, promise: Promise ->
      val auth = requireAuth()
      val cardClient = CardClient(currentContext, auth)

      val card = Card(
        number = cardData.number,
        expirationMonth = cardData.expirationMonth,
        expirationYear = cardData.expirationYear,
        cvv = cardData.cvv,
        postalCode = cardData.postalCode,
        cardholderName = cardData.cardholderName
      )

      cardClient.tokenize(card) { result ->
        when (result) {
          is CardResult.Success -> {
            val nonce = result.nonce
            promise.resolve(mapOf(
              "nonce" to nonce.string,
              "type" to "card",
              "isDefault" to nonce.isDefault,
              "cardNetwork" to nonce.cardType,
              "lastFour" to nonce.lastFour,
              "expirationMonth" to nonce.expirationMonth,
              "expirationYear" to nonce.expirationYear,
              "bin" to nonce.bin
            ))
          }
          is CardResult.Failure -> {
            promise.reject(CodedException("CARD_TOKENIZE_ERROR", result.error.message, result.error))
          }
        }
      }
    }

    // ── Google Pay ─────────────────────────────────────────────────────────

    AsyncFunction("isGooglePayReady") { _: Map<String, Any>, promise: Promise ->
      val auth = requireAuth()
      val client = GooglePayClient(currentContext, auth)
      googlePayClient = client

      client.isReadyToPay(currentContext) { readinessResult ->
        when (readinessResult) {
          is GooglePayReadinessResult.ReadyToPay -> promise.resolve(true)
          else -> promise.resolve(false)
        }
      }
    }

    AsyncFunction("tokenizeGooglePay") { request: GooglePayRequestRecord, promise: Promise ->
      val auth = requireAuth()
      val client = googlePayClient ?: GooglePayClient(currentContext, auth)
      googlePayClient = client

      val launcher = googlePayLauncher
        ?: throw CodedException(
          "GOOGLE_PAY_NOT_READY",
          "GooglePay launcher not initialized. Call initialize() early in your app lifecycle.",
          null
        )

      googlePayPromise = promise

      val totalPriceStatus = when (request.totalPriceStatus) {
        "ESTIMATED" -> GooglePayTotalPriceStatus.TOTAL_PRICE_STATUS_ESTIMATED
        else -> GooglePayTotalPriceStatus.TOTAL_PRICE_STATUS_FINAL
      }

      val googlePayRequest = GooglePayRequest(
        currencyCode = request.currencyCode,
        totalPrice = request.totalPrice,
        totalPriceStatus = totalPriceStatus,
        isBillingAddressRequired = request.billingAddressRequired ?: false,
        isEmailRequired = request.emailRequired ?: false,
        isShippingAddressRequired = request.shippingAddressRequired ?: false,
        allowPrepaidCards = request.allowPrepaidCards ?: true,
        googleMerchantName = request.merchantName,
        countryCode = request.countryCode
      )

      client.createPaymentAuthRequest(googlePayRequest) { paymentAuthRequest ->
        when (paymentAuthRequest) {
          is GooglePayPaymentAuthRequest.ReadyToLaunch -> {
            launcher.launch(paymentAuthRequest)
          }
          is GooglePayPaymentAuthRequest.Failure -> {
            googlePayPromise?.reject(
              CodedException("GOOGLE_PAY_ERROR", paymentAuthRequest.error.message, paymentAuthRequest.error)
            )
            googlePayPromise = null
          }
        }
      }
    }

    // ── PayPal Checkout ───────────────────────────────────────────────────

    AsyncFunction("tokenizePayPalCheckout") { request: PayPalCheckoutRequestRecord, promise: Promise ->
      val auth = requireAuth()
      val returnUrl = appLinkReturnUrl
        ?: throw CodedException("PAYPAL_NOT_CONFIGURED", "Return URL not set. Call setReturnUrl() first.", null)

      val client = PayPalClient(currentContext, auth, returnUrl)
      payPalClient = client
      payPalPromise = promise

      val checkoutRequest = PayPalCheckoutRequest(
        amount = request.amount,
        hasUserLocationConsent = true,
        intent = when (request.intent) {
          "sale" -> PayPalPaymentIntent.SALE
          "order" -> PayPalPaymentIntent.ORDER
          else -> PayPalPaymentIntent.AUTHORIZE
        },
        currencyCode = request.currencyCode ?: "USD",
        isShippingAddressRequired = request.shippingAddressRequired ?: false,
        isShippingAddressEditable = request.shippingAddressEditable ?: false,
        userAction = if (request.userAction == "commit")
          PayPalPaymentUserAction.USER_ACTION_COMMIT
        else
          PayPalPaymentUserAction.USER_ACTION_DEFAULT,
        displayName = request.displayName
      )

      client.createPaymentAuthRequest(currentContext, checkoutRequest) { paymentAuthRequest ->
        when (paymentAuthRequest) {
          is PayPalPaymentAuthRequest.ReadyToLaunch -> {
            val pendingRequest = payPalLauncher.launch(currentActivity, paymentAuthRequest)
            when (pendingRequest) {
              is PayPalPendingRequest.Started -> {
                payPalPendingRequestString = pendingRequest.pendingRequestString
              }
              is PayPalPendingRequest.Failure -> {
                payPalPromise?.reject(
                  CodedException("PAYPAL_LAUNCH_ERROR", pendingRequest.error.message, pendingRequest.error)
                )
                payPalPromise = null
              }
            }
          }
          is PayPalPaymentAuthRequest.Failure -> {
            payPalPromise?.reject(
              CodedException("PAYPAL_ERROR", paymentAuthRequest.error.message, paymentAuthRequest.error)
            )
            payPalPromise = null
          }
        }
      }
    }

    // ── PayPal Vault ──────────────────────────────────────────────────────

    AsyncFunction("tokenizePayPalVault") { request: PayPalVaultRequestRecord, promise: Promise ->
      val auth = requireAuth()
      val returnUrl = appLinkReturnUrl
        ?: throw CodedException("PAYPAL_NOT_CONFIGURED", "Return URL not set. Call setReturnUrl() first.", null)

      val client = PayPalClient(currentContext, auth, returnUrl)
      payPalClient = client
      payPalPromise = promise

      val vaultRequest = PayPalVaultRequest(
        hasUserLocationConsent = true,
        billingAgreementDescription = request.billingAgreementDescription,
        userAuthenticationEmail = request.userAuthenticationEmail,
        isShippingAddressRequired = request.shippingAddressRequired ?: false,
        isShippingAddressEditable = request.shippingAddressEditable ?: false,
        displayName = request.displayName
      )

      client.createPaymentAuthRequest(currentContext, vaultRequest) { paymentAuthRequest ->
        when (paymentAuthRequest) {
          is PayPalPaymentAuthRequest.ReadyToLaunch -> {
            val pendingRequest = payPalLauncher.launch(currentActivity, paymentAuthRequest)
            when (pendingRequest) {
              is PayPalPendingRequest.Started -> {
                payPalPendingRequestString = pendingRequest.pendingRequestString
              }
              is PayPalPendingRequest.Failure -> {
                payPalPromise?.reject(
                  CodedException("PAYPAL_LAUNCH_ERROR", pendingRequest.error.message, pendingRequest.error)
                )
                payPalPromise = null
              }
            }
          }
          is PayPalPaymentAuthRequest.Failure -> {
            payPalPromise?.reject(
              CodedException("PAYPAL_ERROR", paymentAuthRequest.error.message, paymentAuthRequest.error)
            )
            payPalPromise = null
          }
        }
      }
    }

    // ── Venmo ─────────────────────────────────────────────────────────────

    AsyncFunction("tokenizeVenmo") { request: VenmoRequestRecord, promise: Promise ->
      val auth = requireAuth()
      val returnUrl = appLinkReturnUrl
        ?: throw CodedException("VENMO_NOT_CONFIGURED", "Return URL not set. Call setReturnUrl() first.", null)

      val client = VenmoClient(currentContext, auth, returnUrl)
      venmoClient = client
      venmoPromise = promise

      val venmoRequest = VenmoRequest(
        paymentMethodUsage = when (request.paymentMethodUsage) {
          "multiUse" -> VenmoPaymentMethodUsage.MULTI_USE
          else -> VenmoPaymentMethodUsage.SINGLE_USE
        },
        profileId = request.profileId,
        collectCustomerBillingAddress = request.collectCustomerBillingAddress ?: false,
        collectCustomerShippingAddress = request.collectCustomerShippingAddress ?: false
      )

      client.createPaymentAuthRequest(currentContext, venmoRequest) { paymentAuthRequest ->
        when (paymentAuthRequest) {
          is VenmoPaymentAuthRequest.ReadyToLaunch -> {
            val pendingRequest = venmoLauncher.launch(currentActivity, paymentAuthRequest)
            when (pendingRequest) {
              is VenmoPendingRequest.Started -> {
                venmoPendingRequestString = pendingRequest.pendingRequestString
              }
              is VenmoPendingRequest.Failure -> {
                venmoPromise?.reject(
                  CodedException("VENMO_LAUNCH_ERROR", pendingRequest.error.message, pendingRequest.error)
                )
                venmoPromise = null
              }
            }
          }
          is VenmoPaymentAuthRequest.Failure -> {
            venmoPromise?.reject(
              CodedException("VENMO_ERROR", paymentAuthRequest.error.message, paymentAuthRequest.error)
            )
            venmoPromise = null
          }
        }
      }
    }
  }

  // ── Return Handlers ──────────────────────────────────────────────────

  private fun handleGooglePayReturn(paymentAuthResult: GooglePayPaymentAuthResult) {
    val client = googlePayClient ?: return
    val promise = googlePayPromise ?: return
    googlePayPromise = null

    client.tokenize(paymentAuthResult) { result ->
      when (result) {
        is GooglePayResult.Success -> {
          val nonce = result.nonce
          promise.resolve(mapOf(
            "nonce" to nonce.string,
            "type" to "googlePay",
            "isDefault" to nonce.isDefault,
            "email" to nonce.email,
            "cardNetwork" to nonce.cardNetwork,
            "lastFour" to nonce.lastFour,
            "billingAddress" to serializePostalAddress(nonce.billingAddress),
            "shippingAddress" to serializePostalAddress(nonce.shippingAddress)
          ))
        }
        is GooglePayResult.Failure -> {
          promise.reject(CodedException("GOOGLE_PAY_ERROR", result.error.message, result.error))
        }
        is GooglePayResult.Cancel -> {
          promise.reject(CodedException("GOOGLE_PAY_CANCELLED", "User cancelled Google Pay", null))
        }
      }
    }
  }

  private fun handlePayPalReturn(intent: Intent) {
    val pendingStr = payPalPendingRequestString ?: return
    val client = payPalClient ?: return
    val promise = payPalPromise ?: return

    val paymentAuthResult = payPalLauncher.handleReturnToApp(
      PayPalPendingRequest.Started(pendingStr), intent
    )

    when (paymentAuthResult) {
      is PayPalPaymentAuthResult.Success -> {
        payPalPendingRequestString = null
        payPalPromise = null

        client.tokenize(paymentAuthResult) { result ->
          when (result) {
            is PayPalResult.Success -> {
              promise.resolve(serializePayPalNonce(result.nonce))
            }
            is PayPalResult.Failure -> {
              promise.reject(CodedException("PAYPAL_ERROR", result.error.message, result.error))
            }
            is PayPalResult.Cancel -> {
              promise.reject(CodedException("PAYPAL_CANCELLED", "User cancelled PayPal", null))
            }
          }
        }
      }
      is PayPalPaymentAuthResult.Failure -> {
        payPalPendingRequestString = null
        payPalPromise = null
        promise.reject(CodedException("PAYPAL_ERROR", paymentAuthResult.error.message, paymentAuthResult.error))
      }
      is PayPalPaymentAuthResult.NoResult -> {
        // User returned without completing flow, keep waiting
      }
    }
  }

  private fun handleVenmoReturn(intent: Intent) {
    val pendingStr = venmoPendingRequestString ?: return
    val client = venmoClient ?: return
    val promise = venmoPromise ?: return

    val paymentAuthResult = venmoLauncher.handleReturnToApp(
      VenmoPendingRequest.Started(pendingStr), intent
    )

    when (paymentAuthResult) {
      is VenmoPaymentAuthResult.Success -> {
        venmoPendingRequestString = null
        venmoPromise = null

        client.tokenize(paymentAuthResult) { result ->
          when (result) {
            is VenmoResult.Success -> {
              val nonce = result.nonce
              promise.resolve(mapOf(
                "nonce" to nonce.string,
                "type" to "venmo",
                "isDefault" to nonce.isDefault,
                "username" to nonce.username,
                "billingAddress" to serializePostalAddress(nonce.billingAddress),
                "shippingAddress" to serializePostalAddress(nonce.shippingAddress)
              ))
            }
            is VenmoResult.Failure -> {
              promise.reject(CodedException("VENMO_ERROR", result.error.message, result.error))
            }
            is VenmoResult.Cancel -> {
              promise.reject(CodedException("VENMO_CANCELLED", "User cancelled Venmo", null))
            }
          }
        }
      }
      is VenmoPaymentAuthResult.Failure -> {
        venmoPendingRequestString = null
        venmoPromise = null
        promise.reject(CodedException("VENMO_ERROR", paymentAuthResult.error.message, paymentAuthResult.error))
      }
      is VenmoPaymentAuthResult.NoResult -> {
        // User returned without completing flow, keep waiting
      }
    }
  }

  // ── Helpers ───────────────────────────────────────────────────────────

  private fun requireAuth(): String {
    return authorization ?: throw BraintreeNotInitializedException()
  }

  private fun serializePayPalNonce(nonce: PayPalAccountNonce): Map<String, Any?> {
    return mapOf(
      "nonce" to nonce.string,
      "type" to "paypal",
      "isDefault" to nonce.isDefault,
      "email" to nonce.email,
      "firstName" to nonce.firstName,
      "lastName" to nonce.lastName,
      "billingAddress" to serializePostalAddress(nonce.billingAddress),
      "shippingAddress" to serializePostalAddress(nonce.shippingAddress)
    )
  }

  private fun serializePostalAddress(address: PostalAddress?): Map<String, Any?>? {
    if (address == null) return null
    return mapOf(
      "recipientName" to address.recipientName,
      "streetAddress" to address.streetAddress,
      "extendedAddress" to address.extendedAddress,
      "locality" to address.locality,
      "region" to address.region,
      "postalCode" to address.postalCode,
      "countryCodeAlpha2" to address.countryCodeAlpha2
    )
  }
}

// ── Record Types ──────────────────────────────────────────────────────────

class CardDataRecord : Record {
  @Field val number: String = ""
  @Field val expirationMonth: String = ""
  @Field val expirationYear: String = ""
  @Field val cvv: String? = null
  @Field val postalCode: String? = null
  @Field val cardholderName: String? = null
}

class GooglePayRequestRecord : Record {
  @Field val merchantName: String? = null
  @Field val countryCode: String = "US"
  @Field val currencyCode: String = "USD"
  @Field val totalPrice: String = ""
  @Field val totalPriceStatus: String? = "FINAL"
  @Field val allowPrepaidCards: Boolean? = true
  @Field val billingAddressRequired: Boolean? = false
  @Field val emailRequired: Boolean? = false
  @Field val shippingAddressRequired: Boolean? = false
}

class PayPalCheckoutRequestRecord : Record {
  @Field val amount: String = ""
  @Field val currencyCode: String? = "USD"
  @Field val intent: String? = "authorize"
  @Field val userAction: String? = null
  @Field val displayName: String? = null
  @Field val shippingAddressRequired: Boolean? = false
  @Field val shippingAddressEditable: Boolean? = false
}

class PayPalVaultRequestRecord : Record {
  @Field val billingAgreementDescription: String? = null
  @Field val displayName: String? = null
  @Field val userAuthenticationEmail: String? = null
  @Field val shippingAddressRequired: Boolean? = false
  @Field val shippingAddressEditable: Boolean? = false
}

class VenmoRequestRecord : Record {
  @Field val paymentMethodUsage: String = "singleUse"
  @Field val profileId: String? = null
  @Field val displayName: String? = null
  @Field val collectCustomerBillingAddress: Boolean? = false
  @Field val collectCustomerShippingAddress: Boolean? = false
}

// ── Exceptions ────────────────────────────────────────────────────────────

class BraintreeNotInitializedException :
  CodedException("NOT_INITIALIZED", "Braintree client not initialized. Call initialize() first.", null)
